### ISSUE TITLE
IPv4-Only Pods in Dual-Stack

### ISSUE THREAD

**Issue by renyunkang** [projectcalico/calico#9491]
If I've enabled dual-stack networking, how can I configure specific pods to use only IPv4 (single-stack)?

Is there a configuration method to achieve this, or is it currently unsupported?

**Comment by mazdakn** [comment]
@renyunkang Have you tried `'cni.projectcalico.org/ipAddrsNoIpam'` annotation. More information here: https://docs.tigera.io/calico/latest/reference/configure-cni-plugins#requesting-a-specific-ip-address

**Comment by renyunkang** [comment]
> @renyunkang Have you tried `'cni.projectcalico.org/ipAddrsNoIpam'` annotation. More information here: https://docs.tigera.io/calico/latest/reference/configure-cni-plugins#requesting-a-specific-ip-address

Thank you for your response. @mazdakn  "NoIpam" is not the desired outcome. I still want Calico to automatically manage IP addresses (IPAM), but I need a configuration option to specify which namespaces or pods should only use IPv4, instead of both IPv4 and IPv6, even with dual-stack enabled.

**Comment by caseydavenport** [comment]
Yeah, I agree - the `ipAddrsNoIpam` is exceptionally niche and not most for 99.99% of users :grin:

**Comment by caseydavenport** [comment]
I think an annotation to expose configuration of IP families is appropriate. One thing to think about is whether or not we want to unconditionally allow application developers (i.e., people deploying pods) to control which IP families they have access to regardless of cluster configuration. 

Generally for these types of annotations, you want this to be opt-in within the CNI config file itself in some way. Otherwise a cluster admin can configure Calico to behave one way (i.e., requiring dual stack) and then have app devs override that configuration when it's not desirable.

We may want to make this annotation available only if the cluster admin has deferred this capability to application developers - see the existing CNI config "feature_control" section which is used for this sort of thing: https://github.com/projectcalico/calico/blob/840cb655c51f40782cfd49f688d1939e7e050abf/cni-plugin/pkg/types/types.go#L33-L37

**Comment by caseydavenport** [comment]
Would be something like `IPFamilySpecification bool` added to that struct, and then checked before parsing the annotation.

**Comment by caseydavenport** [comment]
On second thought.... maybe that's too paranoid. The `Service` struct already allows application devs to configure how their services work. This seems like a fair parallel.
